File: 0725-split-linked-list-in-parts/0725-split-linked-list-in-parts.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def splitListToParts(self, head: Optional[ListNode], k: int) -> List[Optional[ListNode]]:
        length=0
        curr=head
        while curr:
            length+=1
            curr=curr.next
        length, remainder = length // k, length % k

        parts=[]
        curr=head
        
        while curr:
            parts.append(curr)
            part_len = length
            if remainder:
                part_len+=1
                remainder-=1
            
            while part_len and curr:
                part_len-=1
                prev=curr
                curr=curr.next
            prev.next = None
        
        while len(parts)<k:
            parts.append(None)
        return parts
            

File: 0725-split-linked-list-in-parts/0725-split-linked-list-in-parts.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode[]}
 */

const splitListToParts = (head, k) => {
    // Get total length
    let curr = new ListNode(0, head);
    let length = 0;

    while (curr.next) {
        curr = curr.next;
        length += 1;
    }

    // Get lengths of parts
    const nodes = Math.floor(length / k);
    const extra = length % k;
    const lengths = Array.from({ length: k }, (_, i) => nodes + (i < extra ? 1 : 0));

    const parts = [];
    curr = new ListNode(0, head);

    for (let n of lengths) { // For each part
        let temp = curr; // Remember tail of last part
        parts.push(curr.next); // Add head of part to result
        for (let i = 0; i < n; i++) {
            curr = curr.next; // Move to tail of current part
        }
        temp.next = null; // Cut tail of last part
    }
    return parts;
}
